lib/spack/spack/report.py

```diff
@@ -247,30 +260,196 @@ class collect_info(object):
+                                             'Site.xml')
+                t = env.get_template(site_template)
+                f.write(t.render(report_data))
+
+                phase_template = os.path.join(templates[self.format_name],
+                                              report_name)
+                t = env.get_template(phase_template)
+                f.write(t.render(report_data))
+            self.upload_to_cdash(phase_report)
+
+    def concretization_report(self, msg):
+        if not self.format_name == 'cdash':
+            return
+
+        report_data = {}
+        report_data['starttime'] = self.starttime
+        report_data['endtime'] = self.starttime
+        self.cdash_initialize_report(report_data)
+
+        output_filename = os.path.join(self.filename, 'Update.xml')
+        with open(output_filename, 'w') as f:
+            report_data['msg'] = msg
+            env = spack.tengine.make_environment()
+            update_template = os.path.join(templates[self.format_name],
+                                           'Update.xml')
+            t = env.get_template(update_template)
+            f.write(t.render(report_data))
+        self.upload_to_cdash(output_filename)
+
+    def upload_to_cdash(self, filename):
+        if not self.upload_site:
+            return
+
+        opener = build_opener(HTTPHandler)
+        with open(filename, 'rb') as f:
+            filesize = os.path.getsize(filename)
+
+            # Compute md5 checksum for the contents of this file.
+            m = hashlib.md5()
+            while True:
+                data = f.read(8192)
+                if not data:
+                    break
+                m.update(data)
+            md5sum = m.hexdigest()
+            f.seek(0)
+
+            url = "{0}&MD5={1}".format(self.upload_site, md5sum)
+            request = Request(url, data=f)
+            request.add_header('Content-Type', 'text/xml')
+            request.add_header('Content-Length', filesize)
+            # By default, urllib2 only support GET and POST.
+            # CDash needs expects this file to be uploaded via PUT.
+            request.get_method = lambda: 'PUT'
+            url = opener.open(request)
+
     def __exit__(self, exc_type, exc_val, exc_tb):
         if self.format_name:
             # Close the collector and restore the
             # original PackageBase.do_install
             self.collector.__exit__(exc_type, exc_val, exc_tb)
 
-            # Write the report
-            with open(self.filename, 'w') as f:
-                env = spack.tengine.make_environment()
-                t = env.get_template(templates[self.format_name])
-                f.write(t.render({'test_suites': self.collector.test_suites}))
+            report_data = {'specs': self.collector.specs}
+
+            if self.format_name == 'cdash':
+                # CDash reporting results are split across multiple files.
+                self.cdash_build_report(report_data)
+            else:
```

Comment:
I think it's worth trying to generalize the code here to something like:
```python
if self.format_name:
    # Close the collector and restore the
    # original PackageBase.do_install
    self.collector.__exit__(exc_type, exc_val, exc_tb)

    report_data = {'specs': self.collector.specs}
    templates[self.format_name](report_data)
```
and have `templates` mapping format names to callable writers. In this way we can remove all the CDash specific code from the class that collects the information.